pep-0544.txt

```diff
@@ -0,0 +1,1006 @@
+aliases::
+
+  from typing import TypeVar, Reversible, Iterable, Sized
+
+  T = TypeVar('T')
+  class SizedIterable(Iterable[T], Sized, Protocol):
+      pass
+  CompatReversible = Union[Reversible[T], SizedIterable[T]]
+
+
+.. _discussion:
+
+``@runtime`` decorator and narrowing types by ``isinstance()``
+--------------------------------------------------------------
+
+The default semantics is that ``isinstance()`` and ``issubclass()`` fail
+for protocol types. This is in the spirit of duck typing -- protocols
+basically would be used to model duck typing statically, not explicitly
+at runtime.
+
+However, it should be possible for protocol types to implement custom
+instance and class checks when this makes sense, similar to how ``Iterable``
+and other ABCs in ``collections.abc`` and ``typing`` already do it,
+but this is limited to non-generic and unsubscripted generic protocols
+(``Iterable`` is statically equivalent to ``Iterable[Any]`).
+The ``typing`` module will define a special ``@runtime`` class decorator
+that provides the same semantics for class and instance checks as for
+``collections.abc`` classes, essentially making them "runtime protocols"::
+
+  from typing import runtime, Protocol
+
+  @runtime
+  class Closeable(Protocol):
+      def close(self):
+          ...
+
+  assert isinstance(open('some/file'), Closeable)
+
+Static type checkers will understand ``isinstance(x, Proto)`` and
+``issubclass(C, Proto)`` for protocols defined with this decorator (as they
+already do for ``Iterable`` etc.). Static type checkers will narrow types
+after such checks by the type erased ``Proto`` (i.e. with all variables
+having type ``Any`` and all methods having type ``Callable[..., Any]``).
+Note that ``isinstance(x, Proto[int])`` etc. will always fail in agreement
+with PEP 484. Examples::
+
+  from typing import Iterable, Iterator, Sequence
+
+  def process(items: Iterable[int]) -> None:
+      if isinstance(items, Iterator):
+          # 'items' have type 'Iterator[int]' here
+      elif isinstance(items, Sequence[int]):
+          # Error! Can't use 'isinstance()' with subscripted protocols
+
+Note that instance checks are not 100% reliable statically, this is why
+this behavior is opt-in, see section on `rejected`_ ideas for examples.
+
+
+Using Protocols in Python 2.7 - 3.5
+===================================
+
+Variable annotation syntax was added in Python 3.6, so that the syntax
+for defining protocol variables proposed in `specification`_ section can't
+be used in earlier versions. To define these in earlier versions of Python
+one can use properties::
+
+  class Foo(Protocol):
+      @property
+      def c(self) -> int:
+          return 42         # Default value can be provided for property...
+
+      @abstractproperty
+      def d(self) -> int:   # ... or it can be abstract
+          return 0
+
+In Python 2.7 the function type comments should be used as per PEP 484.
```

Comment:
I think they should also be allowed for Python 3 (even 3.6+) and the text earlier that introduces these should mention this. (It's fine of course to recommend PEP 526 syntax and to use it in all examples except one showing the fallback syntax.)